src/shiver/models/histogram.py

```diff
@@ -76,3 +89,92 @@ def finishHandle(self):  # pylint: disable=invalid-name
     def errorHandle(self, msg):  # pylint: disable=invalid-name
         """Call parent upon algorithm error"""
         self.parent.finish_loading(self, self.filename, self.ws_type, True, msg)
+
+
+class ADSObserver(AnalysisDataServiceObserver):
+    """Object to handle interactions with the ADS"""
+
+    def __init__(self):
+        super().__init__()
+        self.observeClear(True)
+        self.observeAdd(True)
+        self.observeDelete(True)
+        self.observeReplace(True)
+        self.observeRename(True)
+        self.callback = None
+
+    def clearHandle(self):  # pylint: disable=invalid-name
+        """Callback handle for ADS clear"""
+        logger.debug("clearHandle")
+        if self.callback:
+            # NOTE: When closing the application, mantid will clear the ADS after
+            #       Shiver is closed, which will cause a RuntimeError as widgets
+            #       under shiver is no longer in scope.
+            try:
+                self.callback("clear", None, None)
+            except RuntimeError:
+                pass
+
+    def addHandle(self, ws, _):  # pylint: disable=invalid-name
+        """Callback handle for ADS add"""
+        logger.debug(f"addHandle: {ws}")
+        if self.callback:
+            self.callback("add", ws, filter_ws(ws))
+
+    def deleteHandle(self, ws, _):  # pylint: disable=invalid-name
+        """Callback handle for ADS delete"""
+        logger.debug(f"deleteHandle: {ws}")
+        if self.callback:
+            self.callback("del", ws, None)
+
+    def replaceHandle(self, ws, _):  # pylint: disable=invalid-name
+        """Callback handle for ADS replace"""
+        logger.debug(f"replaceHandle: {ws}")
+        if self.callback:
+            self.callback("del", ws, None)
+            self.callback("add", ws, filter_ws(ws))
+
+    def renameHandle(self, old, new):  # pylint: disable=invalid-name
+        """Callback handle for ADS rename"""
+        logger.debug(f"renameHandle: {old} {new}")
+        if self.callback:
+            self.callback("del", old, None)
+            self.callback("add", new, filter_ws(new))
+
+    def register_call_back(self, callback):
+        """Set the callback function for workspace changes"""
+        self.callback = callback
+
+
+def filter_ws(name):
+    """Return the type of workspace"""
+    ws_id = mtd[name].id()
+    ws_type = None
+
+    if ws_id == "MDHistoWorkspace":
+        ws_type = "mdh"
+    elif ws_id == "Workspace2D":
+        # verify if it is one bin per histogram
+        if mtd[name].blocksize() == 1:
+            ws_type = "norm"
+        else:
+            logger.error(f"Workspace2D {name} has more than one bin per histogram")
+    elif ws_id == "MDEventWorkspace<MDEvent,4>":
```

Comment:
This won't match for >4D, maybe check `ws_id.startswith("MDEventWorkspace") and mtd[name].getNumDims() >=4`

Reply:
The condition has been updated following given suggestions.